src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,167 @@
+        for (T element : list) {
+            addLastEntry(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size);
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size);
+        Node<T> setNode = findNodeByIndex(index);
+        T element = setNode.item;
+        setNode.item = value;
+        return element;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        Node<T> removeNode = findNodeByIndex(index);
+        unlink(removeNode);
+        return removeNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> removeNode = head;
+        for (int i = 0; i < size; i++) {
+            if (removeNode.item == object || removeNode.item != null
+                    && removeNode.item.equals(object)) {
+                unlink(removeNode);
+                return true;
+            }
+            removeNode = removeNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index, int size) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index:" + index + ", Size:"
+                                                + size);
+        }
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.previous != null) {
+            node.previous.next = node.next;
+        } else {
+            head = node.next;
+        }
+
+        if (node.next != null) {
+            node.next.previous = node.previous;
+        } else {
+            tail = node.previous;
+        }
+        size--;
+    }
+
+    private void addLastEntry(T element) {
+        Node<T> newLastNode = new Node<>(tail, element, null);
+        if (tail != null) {
+            tail.next = newLastNode;
+        }
+        if (head == null) {
+            head = newLastNode;
+        }
+        tail = newLastNode;
+        size++;
+    }
+
+    private void addFirstEntry(T element) {
+        Node<T> newFirstNode = new Node<>(null, element, head);
+        if (head != null) {
+            head.previous = newFirstNode;
+        }
+        if (tail == null) {
+            tail = newFirstNode;
+        }
+        head = newFirstNode;
+        size++;
+    }
+
+    private void addByIndexEntry(T element, int index) {
+        Node<T> nodeRight = findNodeByIndex(index);
+        Node<T> nodeLeft = new Node<T>(nodeRight.previous, element, nodeRight.next);
+        nodeRight.previous.next = nodeLeft;
+        nodeRight.previous = nodeLeft;
+        size++;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
```

Comment:
include index check here instead of repeatedly calling them together like this:
![image](https://github.com/mate-academy/jv-linked-list/assets/49767049/39fe0eb2-f996-4d3e-9e7e-bf8e0004b199)